### ISSUE TITLE
moto batch - "RuntimeError: cannot join thread before it is started"

### ISSUE THREAD

**Issue by tzven0** [getmoto/moto#7099]
Hey,

**after upgrading from moto 4.0.11 to 4.2.11** i am facing the following issue with my previously running pytest. The first test is successfull, the second test, essentially doing the same but with different input, fails.

This is the **traceback**:

```
========================================================================================== FAILURES ========================================================================================== 
_____________________________________________________________________________ test_2 _____________________________________________________________________________ 

args = ()
kwargs = {'context_sample': '', 'fake_test_2_entry': '<?xml version="1.0" encoding="UTF-8"?>\n<BatchList>\n    <BatchEntry ...quired fields'}, 'pk': {'S': '#step#xyz'}, 'sk': {'S': '#templates'}, 'val': {'L': [{'M': {...}}, {'M': {...}}]}}]}

    def wrapper(*args: "P.args", **kwargs: "P.kwargs") -> T:
        self.start(reset=reset)
        try:
>           result = func(*args, **kwargs)

C:\Users\xxx\AppData\Roaming\Python\Python311\site-packages\moto\core\models.py:150:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  
C:\Users\xxx\AppData\Roaming\Python\Python311\site-packages\moto\core\models.py:150: in wrapper
    result = func(*args, **kwargs)
C:\Users\xxx\AppData\Roaming\Python\Python311\site-packages\moto\core\models.py:148: in wrapper
    self.start(reset=reset)
C:\Users\xxx\AppData\Roaming\Python\Python311\site-packages\moto\core\models.py:113: in start
    backend.reset()
C:\Users\xxx\AppData\Roaming\Python\Python311\site-packages\moto\core\base_backend.py:229: in reset
    region_specific_backend.reset()
C:\Users\xxx\AppData\Roaming\Python\Python311\site-packages\moto\batch\models.py:1058: in reset
    job.join(0.2)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

self = <Job(MOTO-BATCH-bdbc5d83-5429-4c00-b3ed-a4ef9668fba3, initial daemon)>, timeout = 0.2

    def join(self, timeout=None):
        """Wait until the thread terminates.

        This blocks the calling thread until the thread whose join() method is
        called terminates -- either normally or through an unhandled exception
        or until the optional timeout occurs.

        When the timeout argument is present and not None, it should be a
        floating point number specifying a timeout for the operation in seconds
        (or fractions thereof). As join() always returns None, you must call
        is_alive() after join() to decide whether a timeout happened -- if the
        thread is still alive, the join() call timed out.

        When the timeout argument is not present or None, the operation will
        block until the thread terminates.

        A thread can be join()ed many times.

        join() raises a RuntimeError if an attempt is made to join the current
        thread as that would cause a deadlock. It is also an error to join() a
        thread before it has been started and attempts to do so raises the same
        exception.

        """
        if not self._initialized:
            raise RuntimeError("Thread.__init__() not called")
        if not self._started.is_set():
>           raise RuntimeError("cannot join thread before it is started")
E           RuntimeError: cannot join thread before it is started

C:\Program Files\Python311\Lib\threading.py:1107: RuntimeError
```

I am trying the following:

```
@mock_dynamodb
@mock_s3
@mock_batch
@mock_iam
def test_2(context_sample, event, put_data_sample, fake_test_2_entry):
    _create_dynamodb_test_table()
    _test_put_item(put_data_sample)
    _test_s3_put(fake_single_entry_bpl)
    _test_s3_put_custom_cfg("test config content", "test.cfg")
    _test_s3_bucket(os.environ["bucket_name"])
    _test_trigger_invocation(context_sample, event)
```

I am switching back to 4.0.11 for now.

**Comment by bblommers** [comment]
Hi @tzven0, can you share a reproducible test case?

**Comment by bblommers** [comment]
@tzven0 I've found one error scenario where the `submit_job` operation is called with the `arrayProperties` argument, but the test case does not wait for the execution of the job to finish.

```
def test_submit_job_array_size():
    # Setup
    job_definition_name = f"sleep10_{str(uuid4())[0:6]}"
    batch_client = ...
    queue_arn = ...

    # Execute
    resp = batch_client.submit_job(
        jobName="test1",
        jobQueue=queue_arn,
        jobDefinition=job_definition_name,
        arrayProperties={"size": 2},
    )
    return
```
That throws the same error.

Is that the same/similar to your test case?

**Comment by tzven0** [comment]
Hey @bblommers,
yes, our testcase involves array type jobs and therefore arrayProperties with size being set.

**Comment by tzven0** [comment]
Thanks for having a look. i am kind of under water at the moment.

**Comment by bblommers** [comment]
No problem @tzven0! This should be fixed as of moto >= 4.2.12.dev9 - please let us know if you run into any other issues.
